configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java

```diff
@@ -682,6 +682,97 @@ private boolean sessionLifeTimeElapsed(Instant created) {
         return created.plus(sessionLifetime).isBefore(clock.instant());
     }
 
+    public void deleteExpiredRemoteAndLocalSessions(Clock clock, Predicate<Session> sessionIsActiveForApplication) {
+        // All known sessions, both local (file) and remote (zookeeper)
+        Set<Long> sessions = getLocalSessionsIdsFromFileSystem();
+        sessions.addAll(getRemoteSessionsFromZooKeeper());
+        log.log(Level.FINE, () -> "Sessions for tenant " + tenantName + ": " + sessions);
+
+        // Skip sessions newly added (we might have a session in the file system, but not in ZooKeeper,
+        // we will exclude these)
+        Set<Long> newSessions = findNewSessionsInFileSystem();
+        sessions.removeAll(newSessions);
+
+        // Avoid deleting too many in one run
+        int deleteMax = (int) Math.min(1000, Math.max(50, sessions.size() * 0.05));
+        int deleted = 0;
+        for (Long sessionId : sessions) {
+            try {
+                Session session = remoteSessionCache.get(sessionId);
+                if (session == null)
+                    session = new RemoteSession(tenantName, sessionId, createSessionZooKeeperClient(sessionId));
+
+                Session.Status status = session.getStatus();
+                boolean activeForApplication = sessionIsActiveForApplication.test(session);
+                if (status == ACTIVATE && activeForApplication) continue;
+
+                Instant createTime = session.getCreateTime();
+                boolean hasExpired = hasExpired(createTime);
+                if ( ! hasExpired) continue;
+
+                log.log(Level.FINE, () -> "Remote session " + sessionId + " for " + tenantName + " has expired, deleting it");
+                deleteRemoteSessionFromZooKeeper(session);
+                deleted++;
```

Comment:
So I guess this is racy: status and active-session are written in one transaction, but here we read them one after the other, so we could potentially see `PREPARED` and `true`, and then go on to delete the remote session, if it had spent a long time in prepare (not impossible when a session fails to activate because of host provisioning). 
I guess we could handle that by simply re-reading status if `activeForApplication` but `status != ACTIVATE`, or re-reading active-session in the other case? If they still disagree, we should probably warn about it?

Reply:
I wonder if we should also bail out if they disagree, unless the session `isOldAndCanBeDeleted(...)`, like for local sessions?

Reply:
Hmm, we could also read the session status as `PREPARED`, and active as `false`, but then the session is activated while we delete it. I don't remember without looking at the code, but I believe those writes are protected by a lock. Alternatively, we could do the deletion with a zknode version condition ... which I'm not sure we expose through our own APIs >_<

Reply:
Will do this while holding session lock instead